### PR TITLE
feat(plugin): showBadgesInChat

### PR DESCRIPTION
**Pull request by KrstlSkll69**
supersedes the abandoned #1980

Shows profile badges in chat. That includes built in Discord Badges. Also shows Vencord Contributor Badges and all Donor badges (more sources of badges to most likely come soon)

### CODE REVIEW

File: src/plugins/showBadge/index.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    discordNitroBadgePosition: {
+        type: OptionType.NUMBER,
+        description: "The position of the Discord Nitro badge.",
+        default: 3
+    }
+});
+
+export default definePlugin({
+    name: "ShowBadgesInChat",
+    authors: [Devs.Shalev, Devs.fres, Devs.ryan, Devs.KrystalSkull],
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:169]: don't introduce this many authors. only add who really contributed a non trivial amount of code

**Comment by KrstlSkll69** [comment]
@Vendicated ok fixed it, made the "MATCH' more specific

**Comment by KrstlSkll69** [comment]
^ reworked some stuff, fixed typo (i think)

**Comment by Inbestigator** [comment]
KrystalSkull™©® says that they'll say [my pr](https://github.com/Vendicated/Vencord/pull/2239) is amazing if I say theirs is.

# This is a great PR

File: src/plugins/showBadge/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# showBadgesInChat
```
Comment (Vendicated) [inline src/plugins/showBadge/README.md:1]:
```suggestion
# ShowBadgesInChat
```

File: src/plugins/showBadge/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+Shows profile badges in chat. That includes built in Discord Badges. Also shows Vencord Contributor Badges and all Donor badges (more sources of badges to most likely come soon)
+
+-------
+Roadmap:
+ - Include GobalBadges
+ - Just do some optimizing 
+ - Probably a better example photo/more example photos
+-------
```
Comment (Vendicated) [inline src/plugins/showBadge/README.md:9]: don't include a Roadmap in the README. the user doesn't care

you should instead complete this roadmap before merge

File: src/plugins/showBadge/index.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    return isPluginDev(userID) ? [
+        <span style={{ order: settings.store.vencordContributorBadgePosition }}>
+            <RoleIconComponent
+                className={roleIconClassName}
+                name={"Vencord Contributor"}
+                size={20}
+                src={"https://vencord.dev/assets/favicon.png"}
+            />
+        </span>
+    ] : [];
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:71]: why does this return a single element array?

why is this not a proper react component?

File: src/plugins/showBadge/index.tsx [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            {BadgeApi.getDonorBadges(userID)?.map(badge =>
+                <RoleIconComponent
+                    className={roleIconClassName}
+                    name={badge.description}
+                    size={20}
+                    src={badge.image}
+                />
+            )}
+        </span>
+    ];
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:58]: why does this return a single element array?

why is this not a proper react component?

File: src/plugins/showBadge/index.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+function vencordDonorChatBadges(userID: string) {
+    return [
+        <span style={{ order: settings.store.vencordDonorBadgesPosition }}>
+            {BadgeApi.getDonorBadges(userID)?.map(badge =>
+                <RoleIconComponent
+                    className={roleIconClassName}
+                    name={badge.description}
+                    size={20}
+                    src={badge.image}
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:54]: needs a key

File: src/plugins/showBadge/index.tsx [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            badges.push(
+                <RoleIconComponent
+                    className={roleIconClassName}
+                    name={curr[1]}
+                    size={20}
+                    src={`https://cdn.discordapp.com/badge-icons/${curr[2]}.png`}
+                />
+            );
+        return badges;
+    }, []);
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:87]: this reduce is confusing. more logical syntax would be
```js
discordBadges.filter(badge => (userFlags & 1 <<< badge[0])).map(badge => <RoleIconComponent ... />)
```

File: src/plugins/showBadge/index.tsx [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+    return chatBadges.length > 0 ? [
+        <span style={{ order: settings.store.discordProfileBadgesPosition }}>
+            {chatBadges}
+        </span>
+    ] : [];
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:93]: please stop with the single element arrays. use proper react components

File: src/plugins/showBadge/index.tsx [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+function ChatBadges({ author }) {
+    const chatBadges = [
+        ...settings.store.showVencordDonorBadges ? vencordDonorChatBadges(author.id) : [],
+        ...settings.store.showVencordContributorBadges ? vencordContributorChatBadge(author.id) : [],
+        ...settings.store.showDiscordProfileBadges ? discordProfileChatBadges(author.flags || author.publicFlags) : [],
+        ...settings.store.showDiscordNitroBadges ? discordNitroChatBadge(author.premiumType) : []
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:114]: use proper react syntax
```jsx
{settings.store.showVencordDonorBadges && <VencordDonorChatBadge userId={author.id} />}
```

File: src/plugins/showBadge/index.tsx [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+export default definePlugin({
+    name: "ShowBadgesInChat",
+    authors: [Devs.Shalev, Devs.ryan, Devs.KrystalSkull],
+    description: "Shows profile badges in chat. That includes built in Discord Badges. Also shows Vencord Contributor Badges and all Donor badges (more sources of badges to most likely come soon)",
+    dependencies: ["MessageDecorationsAPI"],
+    patches: [
+        {
+            find: "Messages.ROLE_ICON_ALT_TEXT",
+            replacement: {
+                match: /function \i\(\i\){let \i,{className:.*?\)}\)}/,
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:176]: way too vague match. make this more specific

File: src/plugins/showBadge/index.tsx [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                replace: "$self.RoleIconComponent=$&;$&",
+            }
+        }
+    ],
+    settings,
+    set RoleIconComponent(c: any) {
+        RoleIconComponent = c;
+    },
+    start: () => {
+        roleIconClassName = findByProps("roleIcon", "separator").roleIcon;
```
Comment (Vendicated) [inline src/plugins/showBadge/index.tsx:186]: use findByPropsLazy instead of having this inside start

File: src/plugins/showBadge/README.md (hunk 3 above)
Comment (KrstlSkll69) [reply src/plugins/showBadge/README.md]: Already did this, just forgor to delete this part

File: src/plugins/showBadge/README.md (hunk 2 above)
Comment (KrstlSkll69) [reply src/plugins/showBadge/README.md]: done

File: src/plugins/showBadge/index.tsx (hunk 6 above)
Comment (ilyigna) [reply src/plugins/showBadge/index.tsx:54]: Cope.

**Comment by KrstlSkll69** [comment]
closing this to reopen with all the changes made

**Pull request closed without merging**